lib/commands/context.js

```diff
@@ -73,7 +80,121 @@ extensions.getNewRemoteDebugger = async function getNewRemoteDebugger () {
     garbageCollectOnExecute: util.hasValue(this.opts.safariGarbageCollect)
       ? !!this.opts.safariGarbageCollect
       : false,
-  });
+    udid: this.opts.udid,
+  }, this.isRealDevice());
+};
+
+commands.setContext = async function setContext (name, callback, skipReadyCheck) {
+  function alreadyInContext (desired, current) {
+    return (desired === current ||
+           (desired === null && current === NATIVE_WIN) ||
+           (desired === NATIVE_WIN && current === null));
+  }
+
+  log.debug(`Attempting to set context to '${name}'`);
+  if (alreadyInContext(name, this.curContext)) {
+    // already in the named context, no need to do anything
+    return;
+  } else if (name === NATIVE_WIN || name === null) {
+    // switching into the native context
+    this.curContext = null;
+    return;
+  }
+
+  // switching into a webview context
+
+  // if contexts have not already been retrieved, get them
+  if (_.isUndefined(this.contexts)) {
+    await this.getContexts();
+  }
+
+  let contextId = name.replace(WEBVIEW_BASE, '');
+  if (contextId === '') {
+    // allow user to pass in "WEBVIEW" without an index
+    // the second context will be the first webview as
+    // the first is always NATIVE_APP
+    contextId = this.contexts[1];
+  }
+  if (!_.includes(this.contexts, contextId)) {
+    throw new errors.NoSuchContextError();
+  }
+
+  // `contextId` will be in the form of `appId.pageId` in this case
+  let [appIdKey, pageIdKey] = _.map(contextId.split('.'), (id) => parseInt(id, 10));
+  await this.remote.selectPage(appIdKey, pageIdKey, skipReadyCheck);
+  this.curContext = contextId;
+
+  // attempt to start performance logging, if requested
+  if (this.opts.enablePerformanceLogging && this.remote) {
+    log.debug(`Starting performance log on '${this.curContext}'`);
+    this.logs.performance = new IOSPerformanceLog(this.remote);
```

Comment:
should we check for this.logs existence?

Reply:
That would cover up a programming error. If there is no `this.logs` then _we_ have done something wrong.